### PR TITLE
SW Forecast Support - Issue #83

### PR DESCRIPTION
**Pull request by rstoneback**
Initial code for F10.7 cm solar flux and Kp. Includes support for forecast values via the tag name, 'forecast'. F10.7 support includes 'all' data, from 1947 to the present. Perhaps Kp needs an 'all' option as well.

Added instruments to the documentation and __init__.py.

### CODE REVIEW

**Comment by rstoneback** [comment]
Support for Kp, AP, and F10.7 forecasts. The 45-day F10.7 and Ap are in the F10.7 support, tag='45day'. Three day forecasts are in 'forecast' for both F10.7 and Kp instruments. Support also added for F10.7 'all', includes all data available from LASP.

**Comment by rstoneback** [comment]
I need to tweak a line in this pull. Hold off for now.

**Comment by rstoneback** [comment]
Code needs to be updated to modify test dates for forecast data. They are only going to have the current forecast. Timing needs to be robust with respect to the forecast release otherwise we will introduce a time dependent error.

**Comment by rstoneback** [comment]
I also added the tweak referenced previously, but we are back to needing a tweak.

File: pysat/instruments/sw_f107.py [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+platform : string
+    'sw'
+name : string
+    'f107'
+tag : string
+    '' Standard F10.7 data (single day at a time)
+    'all' All F10.7
+    'forecast' Grab forecast data from SWPC (next 3 days)
+    '45day' 45-Day Forecast data frmo the Air Force
+
```
Comment (aburrell) [inline pysat/instruments/sw_f107.py]: typo "from" not "frmo"

**Comment by aburrell** [comment]
Ok, this might be a bit controversial, but it doesn't really make sense to supply a download date for the forecasts.  Can we make the download start and stop times optional, having them default to start=dt.datetime.today(), stop=start+dt.timedelta(days=1)?

**Comment by rstoneback** [comment]
Sure. I don’t see why not.

The easiest way - the user would still have to put in a date but we could just ignore it.

Slightly harder but doable - I could replace the Instrument download as well, for this object only, that doesn’t require dates.

Which is weirder, a function that ignores inputs or one that has a different call structure than all the other downloads?


On Nov 9, 2018, at 8:34 AM, Angeline Burrell <notifications@github.com<mailto:notifications@github.com>> wrote:


Ok, this might be a bit controversial, but it doesn't really make sense to supply a download date for the forecasts. Can we make the download start and stop times optional, having them default to start=dt.datetime.today(), stop=start+dt.timedelta(days=1)?

**Comment by aburrell** [comment]
I think if we just put default values into the instrument downloader that would do the trick.  If someone forgets to put the date in, they download today's file (or get a date error).  That way the behaviour won't change.

Also, the forecast does really ignore whatever date you put in, since it just grabs whatever is available.

**Comment by aburrell** [comment]
I just ran into a problem when loading the 'recent' data.  It will download today's file if I give it the current time, but it won't load it:

```
import pysat
kp = pysat.Instrument('sw', 'kp', 'recent')
stime = dt.datetime.today()
kp.download(stime, stime+dt.timedelta(days=1))
```

> Downloading data to:  /Users/aburrell/Programs/Data/pysatData/sw/kp/recent/
> This routine can only download the current webpage, not archived forecasts
> Updating pysat file list
> pysat is searching for sw kp recent files.
> Found 1 of them.
> Updating instrument object bounds.

```
kp.files.files
``` 
> 2018-11-09    kp_recent_2018-11-09.txt
> dtype: object

`kp.load(date=stime)`

> No sw kp recent data for 11/09/18

```
stime = dt.datetime(2018,11,9)
kp.load(date=stime)
print(kp)
```

> pysat Instrument object
> -----------------------
> Platform: sw
> Name: kp
> Tag: recent
> Satellite id: 
> 
> Data Processing
> ---------------
> Cleaning Level: clean
> Data Padding: None
> Keyword Arguments Passed to load(): {}
> Custom Functions : 
>     No functions applied.
> 
> Orbit Settings
> --------------
> Orbit properties not set.
> 
> Local File Statistics
> ---------------------
> Number of files: 1
> Date Range: 11/09/2018 --- 11/09/2018
> 
> Loaded Data Statistics
> ----------------------
> Date: 11/09/2018
> DOY: 313
> Time range: 10/11/2018 00:00:00 --- 11/09/2018 21:00:00
> Number of Times: 240
> Number of variables: 3
> 
> Variable Names:
> Kp                            high_lat_Kp                   mid_lat_Kp

**Comment by rstoneback** [comment]
That’s a bit strange. I wonder if the datetimes have a different timezone setting or something….
That or there is a bug in basic file/date selection. Perhaps something related to the file list only having one element. I don’t think that specific case is in the Files class testing routines.
Thanks for the notice.

**Comment by jklenzing** [comment]
> I think if we just put default values into the instrument downloader that would do the trick. If someone forgets to put the date in, they download today's file (or get a date error). That way the behaviour won't change. Also, the forecast does really ignore whatever date you put in, since it just grabs whatever is available.

Having a default date for the instrument object is useful.  For the rockets, there is only really one date per mission, so this would be helpful for future improvements.

**Comment by rstoneback** [comment]
Ok. I will sort out a proper automatic download date for the forecast and it’ll probably provide some guidance of a more systematic path forward for the rocket data.

Cheers.

On Nov 13, 2018, at 8:49 AM, Jeff Klenzing <notifications@github.com<mailto:notifications@github.com>> wrote:


I think if we just put default values into the instrument downloader that would do the trick. If someone forgets to put the date in, they download today's file (or get a date error). That way the behaviour won't change. Also, the forecast does really ignore whatever date you put in, since it just grabs whatever is available.

Having a default date for the instrument object is useful. For the rockets, there is only really one date per mission, so this would be helpful for future improvements.

**Comment by rstoneback** [comment]
Let me start with the date thing first.

datetime.today() includes the current time. When pysat receives a date to load, it uses that via slicing on files list to get the correct range of files. If the time (hour, min, sec) is included, it messes that up. Since datetimes are accepted in multiple places, we should probably implement two functions, 
1. A function that strips hour, minute, second, etc from an input datetime and returns one with only year, month, day. We should probably apply this to all methods that accept a datetime.
2. A today helper function attached to the instrument object that gives year, month, day only.

**Comment by rstoneback** [comment]
If I created a today method and a tomorrow helper method I could then use those as defaults for the download method that would make some sense, and it would also sort out the request for default download of the latest data.

**Comment by rstoneback** [comment]
I just pushed a commit that should address the feedback.

Download dates are now optional. Defaults to yesterday to tomorrow. 
Today, yesterday, and tomorrow are now part of the Instrument object. Could be put in utils I suppose.
The dates are filtered on load and download method. Angeline's Kp example now works for me.

To maintain the ability to load previous forecasts and ensure you are loading the latest forecast, I repeat the most recent file with tomorrow's date. Loading the forecast data can be ensured via

    kp = pysat.Instrument('sw', 'kp', tag='recent')
    kp.download()
    kp.load(date=kp.tomorrow())

The best way forward for the rocket data is to ignore the date inputs in the rocket download routine and only download the relevant data, just like the forecast routines.

**Comment by aburrell** [comment]
> The best way forward for the rocket data is to ignore the date inputs in the rocket download routine and only download the relevant data, just like the forecast routines.

Also, when we get the instrument ranges sorted (see #70 ), the instrument download/load routines can check to ensure the input date is within the specified limits and even supply an instrument-specific default (e.g., times from the first instrument paper).

**Pull request merged**